crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -972,24 +973,56 @@ where
 
     // These are the transactions that the other node has, so we have to potentially
     // download them from the remote node.
-    let client = active_authority
-        .net
-        .load()
-        .clone_client(fragment.other.authority());
-    for tx_digest in &fragment.diff.first.items {
-        // TODO: It's possible that by the time we received the proposal we already have some of
-        // the certs. Should do a filter first before trying to download every of them.
-        // The downloading can probably also happen in parallel.
-        let response = client
-            .handle_transaction_info_request(TransactionInfoRequest::from(tx_digest.transaction))
-            .await?;
-        let cert = response
-            .certified_transaction
-            .ok_or(SuiError::CertificateNotfound {
-                certificate_digest: tx_digest.transaction,
-            })?;
-        diff_certs.insert(*tx_digest, cert);
-    }
+    let other_certs = active_authority
+        .state
+        .database
+        .multi_get_certified_transaction(
+            &fragment
+                .diff
+                .first
+                .items
+                .iter()
+                .map(|digests| digests.transaction)
+                .collect::<Vec<_>>(),
+        )?;
+    let (existing, missing): (Vec<_>, Vec<_>) = fragment
+        .diff
+        .first
+        .items
+        .iter()
+        .zip(other_certs)
+        .partition(|(_, cert)| cert.is_some());
+    debug!(
+        existing_size=?existing.len(),
+        missing_size=?missing.len(),
+        "Downloading certificates to augment the fragment"
+    );
+    diff_certs.extend(
+        existing
+            .into_iter()
+            .map(|(digests, cert_opt)| (*digests, cert_opt.unwrap())),
+    );
+
+    let downloads: Vec<_> = missing
+        .into_iter()
+        .map(|(digests, _)| {
+            let active = active_authority.clone();
+            async move {
+                let cert = active
+                    .net
+                    .load()
+                    .handle_cert_info_request(&digests.transaction, Some(Duration::from_secs(5)))
+                    .await?
+                    .certified_transaction
+                    .unwrap();
+                // TODO: Should we insert the cert to the db?
```

Comment:
my experience with node sync is that absolutely yes, we should always save incremental download progress, otherwise you can live lock due to timeouts.

However, I really prefer not to put certs that have not yet been executed into the main cert table, due to the possibility of future bugs in which we inadvertently recover a cert from the previous epoch from the db. AuthorityState::node_sync_store is designed to store such certs. However, you'll want to guarantee that the certs are also enqueued into exec driver to ensure that they get cleaned up.

But, since this is a short term fix, maybe its not worth worrying about all this. Up to you.